kmk/kmk_keyboard.py

```diff
@@ -90,223 +87,354 @@ def __repr__(self):
-            except ImportError:
-                self.hid_helper = AbstractHID
-                print('Bluetooth is unsupported ')
-
-        self._hid_helper_inst = self.hid_helper(**kwargs)
-
-        # Split keyboard Init
-        if self.split_type is not None:
-            try:
-                # Working around https://github.com/adafruit/circuitpython/issues/1769
-                self._hid_helper_inst.create_report([]).send()
-                self.is_target = True
-
-                # Sleep 2s so target portion doesn't "appear" to boot quicker than
-                # dependent portions (which will take ~2s to time out on the HID send)
-                sleep_ms(2000)
-            except OSError:
-                self.is_target = False
-
-            if self.split_flip and not self.is_target:
-                self.col_pins = list(reversed(self.col_pins))
-            if self.target_side == 'Left':
-                self.split_target_left = self.is_target
-            elif self.target_side == 'Right':
-                self.split_target_left = not self.is_target
+                    self.coord_mapping.append(intify_coordinate(ridx, cidx))
+
+    def _init_hid(self):
+        if self.hid_type == HIDModes.NOOP:
+            self._hid_helper = AbstractHID
+        elif self.hid_type == HIDModes.USB:
+            self._hid_helper = USBHID
+        elif self.hid_type == HIDModes.BLE:
+            self._hid_helper = BLEHID
         else:
-            self.is_target = True
+            self._hid_helper = AbstractHID
+        self._hid_helper = self._hid_helper()
 
-        if self.uart_pin is not None:
-            self.uart = self.init_uart(self.uart_pin)
-
-        if self.rgb_pixel_pin:
-            self.pixels = rgb.RGB(self.rgb_config, self.rgb_pixel_pin)
-            self.rgb_config = None  # No longer needed
-            self.pixels.loopcounter = 0
-        else:
-            self.pixels = None
-
-        if self.led_pin:
-            self.led = led.led(self.led_pin, self.led_config)
-            self.led_config = None  # No longer needed
-        else:
-            self.led = None
-
-        self.matrix = self.matrix_scanner(
+    def _init_matrix(self):
+        self.matrix = MatrixScanner(
             cols=self.col_pins,
             rows=self.row_pins,
             diode_orientation=self.diode_orientation,
             rollover_cols_every_rows=getattr(self, 'rollover_cols_every_rows', None),
         )
 
-        # Compile string leader sequences
-        for k, v in self.leader_dictionary.items():
-            if not isinstance(k, tuple):
-                new_key = tuple(KC[c] for c in k)
-                self.leader_dictionary[new_key] = v
+        return self
 
-        for k, v in self.leader_dictionary.items():
-            if not isinstance(k, tuple):
-                del self.leader_dictionary[k]
+    # Only one GC to allow for extentions to have room.
+    # There are random memory allocations without this
```

Comment:
are they random or just fragmented?

Reply:
Clean hard resets on the controller. I believe that I sorted it, but it's only fragmentation if Circuitpython itself has bugs in it. These are clean resets on the controller, not soft resets.

Reply:
ah. let's clean up that comment to be more explicit then about what happens/the risks thereof

Reply:
Let me know if that's clear enough.

Reply:
This actually did not fix that 100% as I believe that there are other issues in Circuitpython. No changes to software, 3 hard reboots, some run into mem aloc issues, so do not. This may as well be removed as it doesn't solve the problem.